Refactoring to facilitate import by third parties

nr-blablacar opened the pull request:
When trying to use logmine as a library, client projects need to use the following import: 
```python
from src.log_mine import LogMine
```
This is not optimal because:
- most users would expect the root import name to be the same (or at least close to) the pip package name. In our case it took us a while to figure it out.
- src is likely to generate name collisions if other packages use the same structure.

I order to improve this situation, I propose this change. It consists in renaming the "src" folder to "logmine_pkg" and updating dependencies.

"logmine_pkg" was chosen because there is already a file named "logmine" at the root of the repository (the tool cli entrypoint actually). Renaming this file would have broken compatibility to previous versions, so we decided to use a "_pkg" suffix for the directory name. At least modern IDEs should suggest to user typing "import logmine".

We are hoping you will accept to review the change and consider to integrate it upsteam.
Best regards

trungdq88 commented:
Hi,

Thanks for this!

Unfortunately, this is an incompatible change, so I can't merge this. I'll keep this in case the new major version is released.

nr-blablacar commented:
Hi,

Thanks for your reply.

> this is an incompatible change

Could you please explain how this change breaks compatibility ? Do you refer to the import interface, meaning that `import src` will not work anymore ?

Actually, I assumed that direct import was not an already existing use-case due to the unusual naming (src). If that is not the case, it is still possible to create a proxy package named src to maintain the previous interface (at the expense of possible name collisions).

If you prefer this approach, please let me know and I will update my PR.

trungdq88 commented:
> Do you refer to the import interface, meaning that import src will not
work anymore ?

Yes. I'm not sure if anyone is using it out there, but it's best not to
break compatibility if possible.

A proxy package may work, can you explain a bit more about this approach?
It's been a while since I work with python.

Thanks for taking the time on this. I really appreciate it.

On Thu, 18 Nov 2021 at 8:52 PM, nr-blablacar ***@***.***>
wrote:

> Hi,
>
> Thanks for your reply.
>
> this is an incompatible change
>
> Could you please explain how this change breaks compatibility ? Do you
> refer to the import interface, meaning that import src will not work
> anymore ?
>
> Actually, I assumed that direct import was not an already existing
> use-case due to the unusual naming (src). If that is not the case, it is
> still possible to create a proxy package named src to maintain the previous
> interface (at the expense of possible name collisions).
>
> If you prefer this approach, please let me know and I will update my PR.
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/trungdq88/logmine/pull/11#issuecomment-972884894>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABAE53MSIYPKRUPGJEOOE4TUMUALRANCNFSM5IHINFMA>
> .
>
-- 
Best regards,
Đinh Quang Trung (Mr.)

nr-blablacar commented:
Hello,
I updated my PR with a proxy package implementation (see 2nd commit).
It works like this: a skeleton folder is created with the previous name. It only contains __init__.py that import important classes from the new path. It acts as a relay for legacy code.

I also added a test case for the legacy import.

When you want to deprecate the old behaviour just remove the src folder and the test case.

trungdq88 commented:
Looks good. Thanks for this.
